File: QuickSelect/quickselect.py
def quickselect(arr, k):
    if not arr:
        return None
    if k < 0 or k >= len(arr):
        return None
    return quickselect_helper(arr, 0, len(arr) - 1, k)

def quickselect_helper(arr, left, right, k):
    if left == right:
        return arr[left]
    pivot_index = partition(arr, left, right)
    if k == pivot_index:
        return arr[k]
    elif k < pivot_index:
        return quickselect_helper(arr, left, pivot_index - 1, k)
    else:
        return quickselect_helper(arr, pivot_index + 1, right, k)

def partition(arr, left, right):
    pivot = arr[left]
    i, j = left, right
    while i < j:
        while i < j and arr[j] >= pivot:
            j -= 1
        arr[i] = arr[j]
        while i < j and arr[i] <= pivot:
            i += 1
        arr[j] = arr[i]
    

File: QuickSelect/quickselect.cpp
#include <vector>
#include <iostream>
using namespace std;

int partition(vector<int> &nums, int left, int right);
int quickSelect(vector<int> &nums, int k);

int quickSelect(vector<int> &nums, int k)
{
    int left = 0, right = nums.size() - 1;
    while (left <= right)
    {
        int pivotIndex = partition(nums, left, right);
        if (pivotIndex == k)
        {
            return nums[k];
        }
        else if (pivotIndex < k)
        {
            left = pivotIndex + 1;
        }
        else
        {
            right = pivotIndex - 1;
        }
    }
    return -1;
}

int partition(vector<int> &nums, int left, int right)
{
    int pivot = nums[right];
    int i = left;
    int j = right - 1;
    while (i <= j)
    {
        if (nums[i] < pivot)
        {
            i++;
        }
        else if (nums[j] >= pivot)
        {
            j--;
        }
        else
        {
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
        }
    }

    nums[right] = nums[i];
    nums[i] = pivot;
    return i;
}

int main()
{
    vector<int> nums = {3, 2, 1, 5, 6, 4};
    int k = 2;
    int result = quickSelect(nums, k);
    cout << result << endl;
    return 0;
}